Merge First time startup into dev

austin2118ace opened the pull request:
- creates default folder structure in user home directory 
- copies stand-in example settings and calibration json files that are included in the wheel

austin2118ace committed:
implement module entrypoint and setup logger

austin2118ace committed:
create bpod-rig.IO package

austin2118ace committed:
MODULE: default_setup
creates the default folder structure in the user's home folder or a user-provided override directory

austin2118ace committed:
import and call the default_setup; log the resultant directory

austin2118ace committed:
empty placeholder "default" settings json file

austin2118ace committed:
empty placeholder "default" calibration json file

austin2118ace committed:
Add entries to pyproject.toml so the non-python .json files will be included in the wheel in the event of setting up an environment completely from PyPi

austin2118ace committed:
Import shutil and the example files directory
Copy the included default settings/calibration json files to the appropriate directories
If desired, only copy if the destination directories are empty

austin2118ace committed:
ignore PERF203 as we want to try/except in the loop, if there are multiple files we want to check the success of each individually
reformat using ruff

austin2118ace committed:
call default copy from the main function

austin2118ace committed:
use explicit import

bimac commented on pyproject.toml:
```diff
@@ -27,10 +27,16 @@ Homepage = "https://github.com/sanworks/bpod-rig"
+# We need to explicitly include examples as it is a reserved name
 
+[tool.setuptools.dynamic]
+version = {attr="bpod_rig.__version__"}
```
when hardcoding the version in `pyproject.toml` instead of using a dynamic field you can use uv's nifty [version management system](https://docs.astral.sh/uv/guides/package/#updating-your-version), e.g. `uv version --bump minor` to increment `0.1.3` to `0.2.0`. Granted, it takes longer to type in the command than to change the two digits manually - it does, however, prevent mistakes

bimac commented on pyproject.toml:
```diff
@@ -27,10 +27,16 @@ Homepage = "https://github.com/sanworks/bpod-rig"
-[tool.setuptools.dynamic]
-version = {attr="bpod_rig.__version__"}
+[tool.setuptools]
+include-package-data = true
```
`include-package-data` is true by default

bimac commented on bpod_rig/__main__.py:
```diff
@@ -0,0 +1,19 @@
+
+from bpod_rig.IO import default_setup
+
+logging.basicConfig(level=logging.DEBUG)
```
I know this is probably temporary, but the log level is typically set by the end-user - I'd keep that out of production code

bimac commented on bpod_rig/IO/default_setup.py:
```diff
@@ -0,0 +1,85 @@
+    Bpod folder
+    structure location
+    :return: (pathlib.Path) Bpod directory path
+    """
```
this comes down to taste, but I find reST docstrings unreadable ;-)

bimac commented on bpod_rig/IO/default_setup.py:
```diff
@@ -0,0 +1,85 @@
+        bpod_folder_path.mkdir(parents=True, exist_ok=True)
+
+    for _dir in DEFAULT_SUBDIRS:
+        _path = bpod_folder_path.joinpath(_dir)
```
The underscore to indicate private variables doesn't really make sense for function variables as they're not accessible from outside the function anyways. I'd only use that for instance variables.

bimac commented on bpod_rig/IO/default_setup.py:
```diff
@@ -0,0 +1,85 @@
+    return bpod_folder_path
+
+
+def copy_default_files(bpod_dir: Path, only_if_empty: bool = True):
```
the `only_if_empty` parameters makes for some rather non-straightforward code  (see below). Perhaps rename it to `overwrite` and adapt the logic accordingly?

bimac commented on bpod_rig/IO/default_setup.py:
```diff
@@ -0,0 +1,85 @@
+    bpod_calibration_dir_contents = list(bpod_calibration_dir.iterdir())
+    bpod_settings_dir_contents = list(bpod_settings_dir.iterdir())
+
+    if (only_if_empty and len(bpod_calibration_dir_contents) == 0) or not only_if_empty:
+        # If only_if_empty is true, the calibration dir must be empty; if only_if_empty
+        # is false, we'll copy no matter what
```
yikes :)

bimac reviewed (changes requested):
some minor nitpicking ;-)

austin2118ace committed:
Update docstrings to numpy and change the pycharm settings to autogen numpy docstrings

austin2118ace committed:
rename vars with underscores

austin2118ace committed:
We're gonna hardcode the version to use the `uv version` commands to increment

austin2118ace committed:
this parameter is true by default so we'll remove setting it again

austin2118ace committed:
Bump version: 0.0.2

bimac reviewed (approved):
the `only-if-empty` thing is still unresolved, otherwise all good!

austin2118ace committed:
make this logic actually understandable

JSNeuroDev committed:
- default_setup: Changed folder structure to support multiple state machines per PC
- default_setup: Moved /Bpod/ to Documents, so that user has easy access in the file explorer

JSNeuroDev committed:
default_setup: Restored docstring

JSNeuroDev committed:
- default_setup: replaced references to USB Port with more generic Machine ID
